tchannel-core/src/main/java/com/uber/tchannel/channels/ChannelManager.java

```diff
@@ -41,47 +39,102 @@
  * entry point for shutting down all active Channels.
  */
 public class ChannelManager {
+    private final Map<SocketAddress, Peer> peers = new Hashtable<>();
+    private String host = "0.0.0.0";
+    private int port = 0;
 
-    private final Map<InetSocketAddress, ChannelId> channelMap = new HashMap<>();
-    private final ChannelGroup channels = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    public Connection findOrNew(SocketAddress address, Bootstrap bootstrap) throws InterruptedException {
+        Peer peer = peers.get(address);
+        if (peer == null) {
+            synchronized (peers) {
```

Comment:
you don't have to lock the new and connect part. Instead of put, maybe change Hashtable to concurrenthashmap, and use putIfAbsent?